xds: Add configuration objects for ExtAuthz, GrpcService and Bootstrap changes for GrpcService

sauravzg opened the pull request:
This commit introduces configuration objects for the external authorization (ExtAuthz) filter and the gRPC service and corresponding translations from XDS proto and Bootstrap. These classes provide a structured, immutable representation of the subset of the configuration defined in the xDS protobuf messages.

This PR should mostly now (hopefully ) be compliant with https://github.com/grpc/proposal/pull/510 but without 
- CallCredentials (since I don't see A97) being implemented yet and would prefer to do it in a followup , we return empty optional)
- TlsCredentials( since it's non trivial to construct a TLS credentials object, we throw an exception)
- LocalCredentials(Java does't support these, we throw an exception)

The main new classes are:
- `ExtAuthzConfig`: Represents the configuration for the `ExtAuthz` filter, including settings for the gRPC service, header mutation rules, and other filter behaviors.
- `GrpcServiceConfig`: Represents the configuration for a gRPC service, including the target URI, credentials, and other settings.
- `HeaderMutationRulesConfig`: Represents the configuration for header mutation rules.
- `ChannelCredsConfig`  and friends: To allow comparison between credential configuration , to allow caching based on creds which'll be needed in followup PRs for authz and proc. 

The relevant sections of the spec are 
- GrpcService: https://github.com/grpc/proposal/pull/510
- ExtAuthz: https://github.com/grpc/proposal/pull/481/files#diff-6bb76a24aa142cc33db9218509688f01b30c8885d2fd8849f164244e68cd54eaR106-R190  

This commit also includes parsers to create these configuration objects from the corresponding protobuf messages, as well as unit tests for the new classes.

sauravzg committed:
feat(xds): Add configuration objects for ExtAuthz and GrpcService

This commit introduces configuration objects for the external authorization (ExtAuthz) filter and the gRPC service it uses. These classes provide a structured, immutable representation of the configuration defined in the xDS protobuf messages.

The main new classes are:
- `ExtAuthzConfig`: Represents the configuration for the `ExtAuthz` filter, including settings for the gRPC service, header mutation rules, and other filter behaviors.
- `GrpcServiceConfig`: Represents the configuration for a gRPC service, including the target URI, credentials, and other settings.
- `HeaderMutationRulesConfig`: Represents the configuration for header mutation rules.

This commit also includes parsers to create these configuration objects from the corresponding protobuf messages, as well as unit tests for the new classes.

ejona86 commented on xds/src/main/java/io/grpc/xds/internal/extauthz/ExtAuthzConfig.java:
```diff
@@ -0,0 +1,250 @@
+public abstract class ExtAuthzConfig {
+
+  /** Creates a new builder for creating {@link ExtAuthzConfig} instances. */
+  public static Builder builder() {
```
Nit: we typically name it `newBuilder()`

kannanjgithub replied:
I notice that for @AutoValue.Builder classes we typically use builder() rather than newBuilder() and this ExtAuthzConfig is an `@AutoValue class`, so it won't be inconsistent in that regard.

ejona86 commented on xds/src/main/java/io/grpc/xds/internal/extauthz/ExtAuthzConfig.java:
```diff
@@ -0,0 +1,250 @@
+        .denyAtDisable(extAuthzProto.getDenyAtDisable().getDefaultValue().getValue());
+
+    if (extAuthzProto.hasFilterEnabled()) {
+      builder.filterEnabled(parsePercent(extAuthzProto.getFilterEnabled().getDefaultValue()));
```
Move XdsRouteConfigureResource.parseFractionMatcher to MatcherParser and use it here.

ejona86 commented on xds/src/main/java/io/grpc/xds/internal/grpcservice/GrpcServiceConfig.java:
```diff
@@ -0,0 +1,308 @@
+      } catch (InvalidProtocolBufferException e) {
+        // TODO(sauravzg): Add unit tests when we have a solution for TLS creds.
+        // This code is as of writing unreachable because all channel credential message
+        // types except TLS are empty messages.
```
I see a `unpack(Xdscredentials.class)` up there. That doesn't look empty.

You don't actually need something non-empty to trigger this. You can actually cause this exception parsing `google.protobuf.Empty`. But I understand you're really saying that you aren't decoding the any if you won't look at any fields inside of the message.

sauravzg replied:
Resolving. I think I was able to reach this in unit tests somehow.

sauravzg committed:
Fixup: Address comments from #12492

sauravzg commented on xds/src/main/java/io/grpc/xds/internal/extauthz/ExtAuthzConfig.java:
```diff
@@ -0,0 +1,250 @@
+
+  private static Pattern parseRegex(String regex, String fieldName) throws ExtAuthzParseException {
+    try {
+      return Pattern.compile(regex);
```
Fixed.

sauravzg commented on xds/src/main/java/io/grpc/xds/internal/extauthz/ExtAuthzConfig.java:
```diff
@@ -0,0 +1,250 @@
+   * @return An {@link ExtAuthzConfig} instance.
+   * @throws ExtAuthzParseException if the proto is invalid or contains unsupported features.
+   */
+  public static ExtAuthzConfig fromProto(ExtAuthz extAuthzProto) throws ExtAuthzParseException {
```
I believe the objects are now independent from proto . Some artifacts of documentation may linger despite my attempts to fix wherever applicable. 

The parsing is now moved into a separate class.

There is still not code using it until the very likely the final PR in the chain , where we construct the actual filter.
But it serves as an implementation of the  business logic for translating xds config to java representations.

I would continue to prefer implementations in a bottom up manner, where we define interfaces and implementation,  evaluate them in isolation and then evaluate their use at callsites if possible.

sauravzg commented on xds/src/main/java/io/grpc/xds/internal/grpcservice/GrpcServiceConfig.java:
```diff
@@ -0,0 +1,308 @@
+
+
+/**
+ * A Java representation of the {@link io.envoyproxy.envoy.config.core.v3.GrpcService} proto,
```
Same as other comment . Resolving this to centralize the configuration.

sauravzg committed:
Fixup: 12492 Split HeaderValueValidationUtils to GrpcService to match the updated requirements

sauravzg committed:
Fixup: 12492 Remove artifact from old  channelcreds implementation and address nits.

sauravzg committed:
Fixup 12492: Use builder instead of newBuilder

sauravzg committed:
Fixup: 12492 Fix callcredentials to not apply instead of erroring out and add test coverage

sauravzg committed:
Fixup 12492: Refactor `allowedGrpcService` to be non optional and fix bug


Makes `allowedGrpcServices` to be a non-optional struct instead of
an `Optional<Map<str,AllowedService>>` since it's
essentially an immuatable hash map, making it preferable to use an empty
instance instead of null.

Change a small bug where we continued instead of return when parsing
bootstrap credentials.

sauravzg committed:
Fixup 12492: Address copilot comments

sauravzg committed:
Fixup 12492: Eliminate bootstrap dependency on grpc

sauravzg commented on xds/src/main/java/io/grpc/xds/client/Bootstrapper.java:
```diff
@@ -205,11 +206,18 @@ public abstract static class BootstrapInfo {
+     * Returns an opaque object containing the parsed configuration.
+     */
+    public abstract Optional<Object> allowedGrpcServices();
+
```
This is intentional to keep the dependency surface small.

sauravzg commented on xds/src/main/java/io/grpc/xds/GrpcBootstrapperImpl.java:
```diff
@@ -155,9 +162,93 @@ private static ChannelCredentials parseChannelCredentials(List<Map<String, ?>> j
+        if (creds == null) {
+          return null;
+        }
+        return ConfiguredChannelCredentials.create(creds, new JsonChannelCredsConfig(type, config));
```
This is working as intended. Same as the other PR.

sauravzg committed:
Fixup 12492: Eliminate GrpcService..Provider classes

sauravzg committed:
Fixup 12492: Improve test coverage for config parser

sauravzg committed:
Fixup 12492: Eliminate cyclic dependencies


- Move the parser objects up to `io.grpc.xds`
- Move some common objects to `io.grpc.xds.client`

sauravzg committed:
Fixup #12492: Address gemini assist comments about runtime exception leak

ejona86 commented on xds/src/main/java/io/grpc/xds/client/AllowedGrpcServices.java:
```diff
@@ -0,0 +1,66 @@
+ */
+@Internal
+@AutoValue
+public abstract class AllowedGrpcServices {
```
This class is grpc-bootstrap-specific, so it should be in `io.grpc.xds`. Ditto for ConfiguredChannelCredentials.

ejona86 commented on xds/src/main/java/io/grpc/xds/client/AllowedGrpcServices.java:
```diff
@@ -0,0 +1,66 @@
+  public abstract static class AllowedGrpcService {
+    public abstract ConfiguredChannelCredentials configuredChannelCredentials();
+
+    public abstract Optional<CallCredentials> callCredentials();
```
I don't think there's any need for CallCredentials to be here. Instead, they should be combined into the ChannelCredentials with CompositeChannelCredentials.

ejona86 commented on xds/src/main/java/io/grpc/xds/GrpcServiceConfigParser.java:
```diff
@@ -0,0 +1,339 @@
+          return Optional.empty();
+      }
+    } catch (InvalidProtocolBufferException e) {
+      throw new GrpcServiceParseException("Failed to parse channel credentials: " + e.getMessage());
```
This discards the stack trace. The protobuf exception should be included as the "cause" of the parse exception via the two-arg constructor. Ditto for call creds.

(It can still make sense to append `e.getMessage()` as you are doing, as we'll end up sending an error string to the control plane and that might have useful info. Or it might be `null`.)

ejona86 reviewed (commented):
I glanced, and noticed a few things. Nothing too serious.
